### ISSUE TITLE
Question about the inference speed for using the model in real-time

### ISSUE THREAD

**Issue by msson** [MSiam/TFSegmentation#20]
Hello,

I am trying to test models which are skipnet-mobilenet and shuffleseg with my own dataset (1024x768 resolution)

I've seen in the paper that those models' can be used in real-time situation cause the inference speed is faster than 10fps on PC.

However, when I run your code for inference with my own data, the speed is about 1.5fps and both models speed are similar even though you mentioned shuffeseg is faster than skipnet-mobilenet here. 

Please give me any advice for using the model in real-time.

Thank you.

**Comment by MSiam** [comment]
Hi, 

We have reported on 360x640 image resolution to compare with ENet and computed the number of floating point operations using the profiler tool in tensorflow. MobileNet-SkipNet and ShuffleNet-SkipNet had GFLOPs is 6.2 and 2 respectively. So ShuffleNet should be faster.

We did record the framerate but we havent reported it yet as we are still working on improving it, on TITAN-X the framerate for ShuffleNet was 143 fps and MobileNet was 141 on the same image resolution. However for an efficient inference there are a couple of tricks you need to do like NCHW optimization, there is also the use of optimize_inference tool for Tensorflow that merges some operations and does batchnorm folding. But even without these when I measured the framerate shufflenet was faster than mobilenet.

If you checked our branch optimize_inference and run ./run.sh you will get a graph.pb, then run ./optimize.sh you will get optimized_graph.pb, if you wanna measure inference we ran the infer_optimize.py. But we're still working on it , as this version doesnt have the NCHW optimization yet. Please let us know if you found any bugs in this version of the code.

**Comment by hellochick** [comment]
Hey @MSiam,
Could you tell me the method you calculate fps?  Does the time include i/o time? Because if I use `time.time()` to calculate, this might include the memory time & i/o time. So, please tell me the method you use to calculate inference time, this might be really helpful to my research, thank you.

**Comment by msson** [comment]
@MSiam Hey,
I don't exactly know how to check the model's inference time. When I test the model, I am able to see this.
![1234](https://user-images.githubusercontent.com/35793913/40409585-5591c6d6-5ea6-11e8-8cd5-664f0af93076.png)
"1.21s/it" is this mean that the model inference speed is 1.21s per frame? If it is, it's too slow that I expected.

**Comment by hellochick** [comment]
@msson , in tf.1.8 we have `tf.profiler` can watch timeline, but i'm not sure what tool does he use.

**Comment by msson** [comment]
Does anyone who can watch shuffleseg's inference speed faster than 1.21s/it?

**Comment by MSiam** [comment]
You have to run in inference mode in the run.sh:
python3 main.py --load_config=fcn8s_shufflenet_test.yaml inference Train FCN8sShuffleNet

This will measure the average running time. We are using TF 1.4.
It will call test_inference method in train/train.py if you want to check how its working.

**Comment by msson** [comment]
@MSiam Thanks for your reply. I tried to run inference mode and get the result below.
![111](https://user-images.githubusercontent.com/35793913/40601620-dbea537c-6290-11e8-9e1c-3f5e4bed9341.png)
After I get this result, I checked train.py 814 lines you see below.
![111](https://user-images.githubusercontent.com/35793913/40601686-1a239950-6291-11e8-9d91-498dad247f54.png)
The exit(1) line is the reason for above result but I do not exactly know how could I check the fps.
(I tried to comment exit(1) line but it shows the error result)
So, I only get graph.pb file without any other information when I run inference mode.

**Comment by MSiam** [comment]
What platform are you running on, are you running on the Jetson TX1? Cause otherwise this is much slower than what we reported.
The optimized version as I mentioned earlier is in optimize_inference branch. 
1- Run ./run.sh which will create a graph.pb
2- Run ./optimize.sh that will generate graph_optimized.pb
3- python infer_optimize.py --graph graph_optimized.pb this should give you the running time.

 143 fps the result we got on TITAN-X and around 16 fps on TX2 (this is without NCHW optimization) on 640x360 resolution. If you need more use NCHW optimization and u can check TensorRT.

**Comment by msson** [comment]
@MSiam 
I got this error when I tried to run optimize.sh after I saved graph.pb.
![2018-05-31 09-58-43](https://user-images.githubusercontent.com/35793913/40755380-3dce04f6-64b9-11e8-9549-cf786aaf3f06.png)

**Comment by msson** [comment]
@MSiam Thanks I successfully checked the running time using graph_optimized.pb

**Comment by zeroAska** [comment]
Hi @msson, if possible, would you like to share with me how do you obtain graph_optimized.pb? Also, are you training shuffleseg from scratch?  Thanks!

**Comment by msson** [comment]
@zeroAska Hi, yes I trained shuffleseg from scratch. And the way I got graph_optimized.pb is that I just simply followed author's mention below. (you should check and download optimize.sh and infer_optimize.py on their branch.)
1- Run ./run.sh which will create a graph.pb
2- Run ./optimize.sh that will generate graph_optimized.pb
3- python infer_optimize.py --graph graph_optimized.pb this should give you the running time.

**Comment by zeroAska** [comment]
@MSiam When you are running the inference on Jetson, how much memory did you use? When I tried to do the inference on jetson with a trained mobilenet model in 512x1024 resolution, I found that ~7.8GB memory is needed, but Jetson has only ~4GB available.

**Comment by harsh-agar** [comment]
Hi @MSiam, I get a inference time of about 60 FPS with image resolution 640x360 on GTX 1080 GPU, when using pre-trained fcn8s_shufflenet model  weights, with master branch (without optimization).

Is this speed normal or slow?

**Comment by MSiam** [comment]
We have reported above the one after optimization this is whats measured to compare with ENet since he mentions in his paper he was also fusing operations and batchnorm folding. It was arnd 143 fps on a TITAN-X. I am not sure hw was it before. Let us know if u faced issues with the optimization. 

Also another diff that ENEt is using torch which by default has CHW order while our initial implementation in TF was HWC which is slower u would have to use the CHW implementation if u want it even faster. However the number of operations should be a more stable way to compare as it doesnt have dependnlencies on the env. Thats what we were reporting in the paper.

**Comment by harsh-agar** [comment]
After using optimize_inference branch I get a speed of around 67.5 FPS which is still quite low on GTX 1080 considering you got 143 FPS.

Anything I can do to improve the speed?

Thanks!

**Comment by MSiam** [comment]
I just realized an issue in the inference time we earlier reported it was 143 fps but on a TITAN-X pascal architecture and I contacted ENet author he is working with Maxwell architecture. We're still trying to get access to a GPU with Maxwell architecture to measure the inference time and finalize the optimized + NCHW implementation. However the reported GFLOPs in the paper is correct no issues with that.

**Comment by bloonz** [comment]
Hello @MSiam,
I've tried to measure the inference speed of ShuffleSeg on Nvidia Jetson TX2 (input image size = 512x512) following these instructions:

1- Run ./run.sh which will create a graph.pb
2- Run ./optimize.sh that will generate graph_optimized.pb
3- python3 infer_optimize.py --graph graph_optimized.pb 

The problem is that when I ran the infer_optimize.py several times, the FPS was different for every run, ranging from 8 FPS to 16 FPS.

Do you have any suggestions as to why that might occur?

Thank you.
